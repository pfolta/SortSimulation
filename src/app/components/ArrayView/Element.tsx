import React from "react";
import Measure from "react-measure";
import styled from "styled-components";

interface ElementProps {
    array: number[];
    index: number;
    maxValue: number;
}

interface StyledElementProps extends ElementProps {
    contentWidth?: number;
}

const StyledElement = styled.div.attrs<StyledElementProps>(({ array, index, maxValue, contentWidth }) => ({
    style: {
        fontSize: computeFontSize(contentWidth) + "px",
        height: (array[index] / maxValue) * 100 + "%"
    }
}))<StyledElementProps>`
    background-color: #4c566a;
    color: white;

    overflow: hidden;

    font-size: 0;
    text-align: center;

    :hover {
        background-color: #ebcb8b;
    }
`;

const computeFontSize = (width?: number): number => {
    const size = width ? width / 2 : 0;
    return size < 8 ? 0 : size;
};

const Element = (props: ElementProps): JSX.Element => (
    <Measure>
        {({ contentRect, measureRef }) => (
            <StyledElement ref={measureRef} {...props} contentWidth={contentRect.entry?.width}>
                {props.array[props.index]}
            </StyledElement>
        )}
    </Measure>
);

export default Element;
export { StyledElement };
export { ElementProps };
