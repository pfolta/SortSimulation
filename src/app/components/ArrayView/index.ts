export { default as ArrayView, ArrayViewProps } from "@/app/components/ArrayView/ArrayView";
export { default as Element, ElementProps, StyledElement } from "@/app/components/ArrayView/Element";
