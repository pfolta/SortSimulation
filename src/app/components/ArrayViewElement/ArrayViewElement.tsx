import styled from "styled-components";

interface ArrayViewElementProps {
    value: number;
}

const ArrayViewElement = styled.div.withConfig({
    shouldForwardProp: (prop) => ["children", "style"].includes(prop)
})<ArrayViewElementProps>`
    background-color: #4c566a;
    color: white;

    transition: transform 100ms ease-out;

    :hover {
        transform: translateY(-1rem);

        /* Prevent jitter on hover, see https://codepen.io/csilverman/post/fixing-the-jitter-bug */
        ::before {
            content: "";

            position: absolute;
            top: 0;
            left: 0;
            right: 0;
            bottom: -1rem;

            z-index: -1;
        }
    }
`;

export default ArrayViewElement;
export { ArrayViewElementProps };
