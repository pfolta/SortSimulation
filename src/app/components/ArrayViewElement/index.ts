export type { ArrayViewElementProps } from "@/app/components/ArrayViewElement/ArrayViewElement";
export { default as ArrayViewElement } from "@/app/components/ArrayViewElement/ArrayViewElement";

export type { ImageElementProps } from "@/app/components/ArrayViewElement/ImageElement";
export { default as ImageElement, StyledImageElement } from "@/app/components/ArrayViewElement/ImageElement";

export type { NumberElementProps } from "@/app/components/ArrayViewElement/NumberElement";
export { default as NumberElement, StyledNumberElement } from "@/app/components/ArrayViewElement/NumberElement";
