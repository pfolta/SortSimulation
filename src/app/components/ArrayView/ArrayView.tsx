import React, { useMemo } from "react";
import styled from "styled-components";

import { Element } from "@/app/components/ArrayView";

interface ArrayViewProps {
    array: number[];
}

const Wrapper = styled.div<ArrayViewProps>`
    display: grid;
    grid-template-columns: repeat(${(props) => props.array.length}, 1fr);
    column-gap: 1px;
    align-items: end;

    height: 30vh;
`;

const ArrayView = ({ array }: ArrayViewProps): JSX.Element => {
    const maxValue = useMemo(() => Math.max(...array), [array]);

    return (
        <Wrapper array={array}>
            {array.map((_, index) => (
                <Element key={index} array={array} index={index} maxValue={maxValue} />
            ))}
        </Wrapper>
    );
};

export default ArrayView;
export { ArrayViewProps };
