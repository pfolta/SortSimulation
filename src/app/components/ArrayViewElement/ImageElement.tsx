import React from "react";
import styled from "styled-components";

import { ArrayViewElement, ArrayViewElementProps } from "@/app/components/ArrayViewElement";

interface ImageElementProps extends ArrayViewElementProps {
    image: string;
}

const StyledImageElement = styled(ArrayViewElement)<ImageElementProps>`
    background-image: url("${({ image }) => image}");
    background-position: ${({ array, index }) => ((array[index] - 1) / (array.length - 1)) * 100}% 0%;
    background-repeat: no-repeat;
    background-size: ${({ array }) => array.length * 100}% 100%;

    height: 100%;
`;

const ImageElement = (props: ImageElementProps): JSX.Element => <StyledImageElement {...props} />;

export default ImageElement;
export { ImageElementProps, StyledImageElement };
