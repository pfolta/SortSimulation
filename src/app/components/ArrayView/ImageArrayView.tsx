import React, { memo } from "react";

import { ArrayView, ArrayViewProps } from "@/app/components/ArrayView";
import { ImageElement } from "@/app/components/ArrayViewElement";

interface ImageArrayViewProps extends ArrayViewProps {
    grayscaleMap: boolean[];
    image: string;
}

const ImageArrayView = ({ array, grayscaleMap, image }: ImageArrayViewProps): JSX.Element => (
    <ArrayView array={array}>
        {array.map((value, index) => (
            <ImageElement key={index} value={value} size={array.length} image={image} grayscale={grayscaleMap[index]} />
        ))}
    </ArrayView>
);

export default memo(ImageArrayView);
export { ImageArrayViewProps };
