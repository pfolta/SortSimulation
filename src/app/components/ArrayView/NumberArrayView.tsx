import React, { memo } from "react";
import styled from "styled-components";

import { ArrayView, ArrayViewProps } from "@/app/components/ArrayView";
import { NumberElement } from "@/app/components/ArrayViewElement";

const StyledNumberArrayView = styled(ArrayView)`
    column-gap: 1px;
`;

const NumberArrayView = ({ array }: ArrayViewProps): JSX.Element => {
    const maxValue = Math.max(...array);

    return (
        <StyledNumberArrayView array={array}>
            {array.map((value, index) => (
                <NumberElement key={index} value={value} maxValue={maxValue} />
            ))}
        </StyledNumberArrayView>
    );
};

export default memo(NumberArrayView);
