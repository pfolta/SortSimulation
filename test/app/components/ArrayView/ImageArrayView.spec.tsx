import React from "react";
import { create } from "react-test-renderer";

import { ImageArrayView } from "@/app/components/ArrayView";
import { ImageElement } from "@/app/components/ArrayViewElement";

describe("ImageArrayView", () => {
    const array = [4, 2, 5, 1, 3];
    const image = "/path/to/image.jpg";

    it("renders correctly", () => {
        const component = create(<ImageArrayView array={array} image={image} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders an ImageElement for every array element in the correct order", () => {
        const component = create(<ImageArrayView array={array} image={image} />);
        const arrayElements = component.root.findAllByType(ImageElement);

        expect(arrayElements).toHaveLength(array.length);

        arrayElements.forEach((bar, index) => {
            expect(bar.props.array).toBe(array);
            expect(bar.props.index).toBe(index);
        });
    });
});
