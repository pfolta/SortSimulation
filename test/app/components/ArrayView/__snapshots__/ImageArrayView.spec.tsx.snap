// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageArrayView renders correctly 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(5,1fr);
  -webkit-column-gap: 1px;
  column-gap: 1px;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  height: 30vh;
}

.c1 {
  background-color: #4c566a;
  color: white;
  -webkit-transition: -webkit-transform 100ms ease-out;
  -webkit-transition: transform 100ms ease-out;
  transition: transform 100ms ease-out;
}

.c1:hover {
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
}

.c1:hover::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: -1rem;
  z-index: -1;
}

.c2 {
  background-image: url("/path/to/image.jpg");
  background-position: 75% 0%;
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

.c3 {
  background-image: url("/path/to/image.jpg");
  background-position: 25% 0%;
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

.c4 {
  background-image: url("/path/to/image.jpg");
  background-position: 100% 0%;
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

.c5 {
  background-image: url("/path/to/image.jpg");
  background-position: 0% 0%;
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

.c6 {
  background-image: url("/path/to/image.jpg");
  background-position: 50% 0%;
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

<div
  className="c0"
>
  <div
    className="c1 c2"
  />
  <div
    className="c1 c3"
  />
  <div
    className="c1 c4"
  />
  <div
    className="c1 c5"
  />
  <div
    className="c1 c6"
  />
</div>
`;
